### PR TITLE
Resolve empty statement warning when using PYBIND11_OVERLOAD_PURE_NAME and PYBIND11_OVERLOAD_PURE

### LINKED ISSUE #2166: Redundant semicolon in the definition of PYBIND11_OVERLOAD_PURE_NAME
When using PYBIND11_OVERLOAD_PURE_NAME, like
```
PYBIND11_OVERLOAD_PURE_NAME(std::string, Animal, "go", Go, n_times);
```
The compiler gives a warning "Empty statement".
This would not affect the correctness of the programs, but I found that there is a redundant semicolon in the definition of PYBIND11_OVERLOAD_PURE_NAME:
```
//pybind11.h
#define PYBIND11_OVERLOAD_PURE_NAME(ret_type, cname, name, fn, ...) \
    PYBIND11_OVERLOAD_INT(PYBIND11_TYPE(ret_type), PYBIND11_TYPE(cname), name, __VA_ARGS__) \
    pybind11::pybind11_fail("Tried to call pure virtual function \"" PYBIND11_STRINGIFY(cname) "::" name "\"");
```
The semicolon in the last line can be removed.

### PR DESCRIPTION
**Pull request by YannickJadoul**
Closes #2166

### CODE REVIEW

File: include/pybind11/pybind11.h
Comment (virtuald) [inline include/pybind11/pybind11.h:2175]: The prior version of this macro ended with a `;`, could break some users?

File: include/pybind11/pybind11.h
Comment (virtuald) [inline include/pybind11/pybind11.h:2141]: Same here

**Comment by bstaletic** [comment]
After a bit of sleep, I think the potential breaking change isn't worth it. The resolution to #2166 could as well be "remove the semicolon in user code".

**Comment by YannickJadoul** [comment]
> After a bit of sleep, I think the potential breaking change isn't worth it. The resolution to #2166 could as well be "remove the semicolon in user code".

What made me remove it (rather than comments "well, then don't put the semicolon"), is that there's also a semicolon in the docs?

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [reply include/pybind11/pybind11.h:2175]: Yeah, removing these semicolons is the whole point of the PR, cfr. #2166!

But ... even if some user code breaks, in principle, 1. it's a loud breaking change (it won't silently crash, but it'll just complain during compilation), 2. usage is like that in all the docs and tests, so arguably, using the macro without `;` is undocumented behavior.

**Comment by bstaletic** [comment]
We can change the docs. Also #2166 isn't an error, just a warning raised by turning the `-W` settings up to 11.

**Comment by virtuald** [comment]
Duplicate of https://github.com/pybind/pybind11/pull/2188

**Comment by YannickJadoul** [comment]
> Duplicate-ish of #2188

Ah, I hadn't noticed. Just trying to fix/close some issues.

**Comment by wjakob** [comment]
As long as we don't violate SemVer (i.e. documented behavior no longer compiling), I am okay with this change. It seems to me that there is no such violation (I only had a brief look though).

File: include/pybind11/pybind11.h
Comment (bstaletic) [inline include/pybind11/pybind11.h:2181]: > Is that a good change, the `do { ... } while (false)`?

Assuming we want to make this kind of a breaking change, sure.

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:88]: I don't know where that magic number originated either, so yes, this is a weird piece of code, but it was there already.

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [reply include/pybind11/pybind11.h:2089]: I thought it was. If I understand correctly, this check whether there is a Python override with that name. If the Python override is found, it could be an override done from pybind11, and be a `cpp_function`, so rather than wrapping a wrapper, the wrapper is unwrapped.

Can it still be overloaded? Yes, but I thínk that `cpp_function` itself is just the Python object and takes care of actual overloading by having multiple reference to C++ functions inside. It's `cpp_function::dispatcher` that does actually chooses the C++ overload; so 1 function from the Python `getattr` side.

So, I do think this is getting the override and not the overload?

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [reply include/pybind11/pybind11.h:2074]: Where exactly? I don't find it.

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [reply include/pybind11/pybind11.h:2181]: Is that ever going to break something that wasn't broken before, is the main question?
I can only imagine `if (...) PYBIND11_OVERRIDE(...);` to actually get fixed, rather than broken?

File: include/pybind11/detail/internals.h
Comment (bstaletic) [reply include/pybind11/detail/internals.h:88]: The magic number originates from Boost's `hash_combine()`.

https://stackoverflow.com/questions/35985960/c-why-is-boosthash-combine-the-best-way-to-combine-hash-values

However, that magic number isn't what makes this code rely on `std::hash`'s ABI. It's that we're creating a temporary `std::hash<T>` just to call the objects `operator()(T)`. Were we to use `absl::Hash` we would have to store the hasher object somewhere.

**Comment by YannickJadoul** [comment]
Right, so, I wanted to first still get the docs straight, but Boris is too fast for me :-)
I was planning to finish this PR and then give 3 options:

1. We don't merge this one, but merge #2188, and tell users to turn down their warnings (or even better, include pybind11 as system headers).
2. We just merge the first commit (9230ac4fe78dedbcdcf9af065909170094c72e30), fixing #2166. It would be breakish, but not according to the docs/SemVer (cfr. https://github.com/pybind/pybind11/pull/2325#issuecomment-663979159).
3. We merge this full PR, deprecating the old macros and functions, and finally getting the C++ terminology correct.

Your votes are welcome! :-)

My personal favourite is 3, obviously. Overloading is different from overriding. And it only gets worse, because we do have `py::overload_cast` and pybind11's overloading mechanism on functions. It can also get confusing when users try to mix: e.g. #2351.

To reply to @bstaletic:

> Kinds of breaking changes in this PR:
> 
>     * Moving functions from `pybind11` to `pybind11::detail` namespace.

Yes. As far as I know, just one (`get_type_overload`/`detail::get_type_override`), right?), undocumented, taking a parameter with a type in the `detail` namespace (`detail::type_info`), dealing with the internals/being an auxiliary function to a public-facing templated interface (`get_overload`/`get_override`), and I have created a deprecated alias outside of `detail`. If necessary, I can move it back out of `detail`. I just seemed like an oversight.

>     * Renaming things.

Yes. Though gently doing so, because I created deprecated aliases? Users' code should break. They should just get deprecation warnings.

> I'm not strictly against these changes, however we need them explicitly enumerated in the upgrade guide.

Agreed. Planning on adding a big note in the docs :-)

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:88]: OK. If you wish, it can be a different PR to fix that?

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [reply include/pybind11/pybind11.h:2074]: No problem!

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [reply include/pybind11/pybind11.h:2181]: Oh, yes. I still need to fix that, such that the original macros are the same as they were. That was the whole point of this exercise!

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:88]: (Still interesting, though :-) )

**Comment by bstaletic** [comment]
You forgot to update the `docs/upgrade.rst`. It's the actual "upgrade guide".

**Comment by YannickJadoul** [comment]
Alright, I'm waiting for #2370 before adding to the changelog and upgrade guide to avoid merge conflicts, but I believe this is ready for review!

To reiterate:

>    1. We don't merge this one, but merge #2188, and tell users to turn down their warnings (or even better, include pybind11 as system headers).
> 
>    2. We just merge the first commit ([9230ac4](https://github.com/pybind/pybind11/commit/9230ac4fe78dedbcdcf9af065909170094c72e30)), fixing #2166. It would be breakish, but not according to the docs/SemVer (cfr. [#2325 (comment)](https://github.com/pybind/pybind11/pull/2325#issuecomment-663979159)).
> 
>    3. We merge this full PR, deprecating the old macros and functions, and finally getting the C++ terminology correct.
> 
> 
> Your votes are welcome! :-)
> 
> My personal favourite is 3, obviously. Overloading is different from overriding. And it only gets worse, because we do have `py::overload_cast` and pybind11's overloading mechanism on functions. It can also get confusing when users try to mix: e.g. #2351.

**Comment by rwgk** [comment]
My honest opinion: it is disproportionally disruptive to rename the macros, just to be semantically more correct. If we add a footnote to the documentation, explaining that override would be a more fitting name, but for historical reasons overload stuck, we're good.

My recommendation is to simply remove the semicolon at the end of the `PYBIND11_OVERLOAD_PURE_NAME` macro (as suggested under #2166) and mention it in the release notes. It's very easy to adjust to, therefore I think it's fine doing this in a patch release.

**Comment by YannickJadoul** [comment]
> My honest opinion: it is disproportionally disruptive to rename the macros, just to be semantically more correct. If we add a footnote to the documentation, explaining that override would be a more fitting name, but for historical reasons overload stuck, we're good.

Yes, true, but then following that reasoning, such a mistake can never get fixed?
On the note of disruptive: users will only get a deprecation warning, though, which can be ignored? Everything will keep on working (until some pybind11 3.0.0 or so).

> My recommendation is to simply remove the semicolon at the end of the PYBIND11_OVERLOAD_PURE_NAME macro (as suggested under #2166) and mention it in the release notes. It's very easy to adjust to, therefore I think it's fine doing this in a patch release.

But yes, this is fine with me as well, to just cherry-pick the first commit. Though @bstaletic was arguing that that would be more "breaking" than the deprecating PR...

**Comment by rwgk** [comment]
> > My honest opinion: it is disproportionally disruptive to rename the macros, just to be semantically more correct. If we add a footnote to the documentation, explaining that override would be a more fitting name, but for historical reasons overload stuck, we're good.
> 
> Yes, true, but then following that reasoning, such a mistake can never get fixed?
> On the note of disruptive: users will only get a deprecation warning, though, which can be ignored? Everything will keep on working (until some pybind11 3.0.0 or so).
> 

https://github.com/search?q=PYBIND11_OVERLOAD
I get 12k hits for "code".
I think it'll take at least a couple years until the old name is 90% flushed out of the global code bases. During that time there will be several hundred confused users, seeing the old name in some places, the new in others.
I agree it's a bit sad that this cannot easily be fixed, but I think the cost is too high. The old name is only "puzzling" but I don't think it rises to "misleading".
Please take my opinion only as one vote. If a majority supports the change I'll support it, too!

> > My recommendation is to simply remove the semicolon at the end of the PYBIND11_OVERLOAD_PURE_NAME macro (as suggested under #2166) and mention it in the release notes. It's very easy to adjust to, therefore I think it's fine doing this in a patch release.
> 
> But yes, this is fine with me as well, to just cherry-pick the first commit. Though @bstaletic was arguing that that would be more "breaking" than the deprecating PR...

I believe that'll trickle through the code bases much faster and decrease confusion monotonically, because the macros will be consistent.

**Comment by YannickJadoul** [comment]
> https://github.com/search?q=PYBIND11_OVERLOAD
> I get 12k hits for "code".

Alright, good point! Maybe we can provide a quick search-and-replace script to the upgrade guide.

At any rate, yes, the work is done, so now it's choosing an option. Probably up to @wjakob to decide, but let's see if we get more votes!

**Comment by wjakob** [comment]
I took another look.

I really don't mind the name change, especially given that the previous naming scheme was arguably confusing/incorrect. It's also a good opportunity to clean up the inconsistency regarding semicolon usage.

What I don't like:

1. This adds too much code, by a large margin. It should be possible to re-implement both variants of macros and functions with minimal duplication.

2. I find it too early to annoy users with deprecation warnings, especially given the minor nature of the change. It's not like keeping around both variants is very costly in terms of the underlying implementation (c.f. point 1). My suggestion: let's only explain the new version in the docs, along with an explanation of the change in the upgrade guide, along with a note saying that the old version may be deprecated at some unspecified time.

**Comment by YannickJadoul** [comment]
Right, I'll have another look with that feedback in mind! :-)

**Comment by YannickJadoul** [comment]
(not sure why, but GitHub doesn't allow me to click the "re-request review" for you, @virtuald. But if you're interested, please do have a look :-) )

**Comment by virtuald** [comment]
Changing the names seems like a reasonable compromise to me.

**Review by wjakob — APPROVED** [review]
Looks great now, approved by me!

**Review by bstaletic — APPROVED** [review]
Looks fine to me.

**Comment by YannickJadoul** [comment]
Thanks all! Let's see how users react, on whether we need to update our upgrade guide further, or not :-)

**Pull request merged**
